src/engine/Result.cpp

```diff
@@ -144,118 +166,104 @@ void Result::enforceLimitOffset(const LimitOffsetClause& limitOffset) {
     return std::ranges::all_of(map, [&](const auto& varAndCol) {
       const auto& [columnIndex, mightContainUndef] = varAndCol.second;
-      bool hasUndefined =
-          datatypeCountsPerColumn.at(columnIndex)
-              .at(static_cast<size_t>(Datatype::Undefined)) != 0;
-      return mightContainUndef == ColumnIndexAndTypeInfo::PossiblyUndefined ||
-             !hasUndefined;
+      if (mightContainUndef == ColumnIndexAndTypeInfo::AlwaysDefined) {
+        return std::ranges::all_of(idTable.getColumn(columnIndex), [](Id id) {
+          return id.getDatatype() != Datatype::Undefined;
+        });
+      }
+      return true;
     });
   };
-  if (isDataEvaluated()) {
+  if (isFullyMaterialized()) {
     AD_EXPENSIVE_CHECK(performCheck(varColMap, std::get<IdTable>(data_)));
   } else {
     auto generator = [](cppcoro::generator<IdTable> original,
                         VariableToColumnMap varColMap,
                         auto performCheck) -> cppcoro::generator<IdTable> {
-      bool first = true;
-      for (auto&& idTable : original) {
-        if (first) {
-          first = false;
-          // No need to check subsequent idTables assuming the datatypes
-          // don't change mid result.
-          AD_EXPENSIVE_CHECK(performCheck(varColMap, idTable));
-        }
-        co_yield std::move(idTable);
+      for (IdTable& idTable : original) {
+        // No need to check subsequent idTables assuming the datatypes
+        // don't change mid result.
+        AD_EXPENSIVE_CHECK(performCheck(varColMap, idTable));
+        co_yield idTable;
       }
     }(std::move(idTables()), varColMap, std::move(performCheck));
-    data_ = std::move(generator);
+    data_.emplace<GenContainer>(std::move(generator));
   }
 }
 
 // _____________________________________________________________________________
 void Result::runOnNewChunkComputed(
-    std::function<void(const IdTable&, std::chrono::milliseconds)> function) {
-  AD_CONTRACT_CHECK(!isDataEvaluated());
-  auto generator =
-      [](cppcoro::generator<IdTable> original,
-         std::function<void(const IdTable&, std::chrono::milliseconds)>
-             function) -> cppcoro::generator<IdTable> {
-    ad_utility::timer::Timer timer{ad_utility::timer::Timer::Started};
-    for (auto&& idTable : original) {
-      function(idTable, timer.msecs());
-      co_yield std::move(idTable);
-      timer.start();
+    std::function<void(const IdTable&, std::chrono::microseconds)> onNewChunk,
+    std::function<void(bool)> onGeneratorFinished) {
+  AD_CONTRACT_CHECK(!isFullyMaterialized());
+  auto generator = [](cppcoro::generator<IdTable> original, auto onNewChunk,
+                      auto onGeneratorFinished) -> cppcoro::generator<IdTable> {
+    // Call this within destructor to make sure it is also called when an
+    // operation stops iterating before reaching the end.
+    absl::Cleanup cleanup{
+        [&onGeneratorFinished]() { onGeneratorFinished(false); }};
```

Comment:
I have abstractions (OnDestructionDontThrowDuringStackUnwiniding) that are a little bit safer.
I wonder that this compiles, because I don't think we can assume that `onGeneratorFinished` never throws, at least not statically.